### PR TITLE
Feature/fix overlap warning

### PR DESCRIPTION
**Pull request by kmtsui**
~~Prerequisite: Developed with geant4.10.05.p01 and root_v6.22.06, need to be tested with older (official) software versions.~~

Fix overlapping warnings with WCTE geometry (nuPRISMBeamTest_16cShort_mPMT)
- mPMT related: addition of reflector causes the daughter logical volume to overflow the mother volume. Some dimension parameters are changed accordingly.
- Blacksheet related: 
   1. Some combination of detector dimensions will cause the cap construction out of z-order. Impose boundary condition to avoid such situation.
   2. Blacksheet in the barrel region close to the cap is missing. Extend the blacksheet from the top/bottom most ring to recover the missing part.

### CODE REVIEW

File: include/GdNeutronHPCapture.hh [hunk 1]
```diff
@@ -20,6 +20,14 @@
 #include "G4NeutronHPChannel.hh"
 #include "G4HadronicInteraction.hh"
 
+#include "G4SystemOfUnits.hh"
```
Comment (pdeperio) [inline include/GdNeutronHPCapture.hh]: Why modifying neutron capture files (also below)?

File: src/WCSimRunAction.cc [hunk 2]
```diff
@@ -181,34 +181,34 @@ void WCSimRunAction::BeginOfRunAction(const G4Run* /*aRun*/)
...
-  geomTree->Branch("PMTradius_ID",&pmt_radius_id,"PMTradius_ID/Double_t");
-  geomTree->Branch("PMTtype_OD",pmt_od_string,"PMTtype_OD[50]/Char_t");
-  geomTree->Branch("PMTradius_OD",&pmt_radius_od,"PMTradius_OD/Double_t");
-  geomTree->Branch("numPMT_ID",&numPMT_id,"numPMT_ID/Int_t");
-  geomTree->Branch("numPMT_OD",&numPMT_od,"numPMT_OD/Int_t");
-  geomTree->Branch("Orientation",&orient,"Orientation/Int_t");
-  geomTree->Branch("Offset_x",&offset_x,"Offset_x/Double_t");
-  geomTree->Branch("Offset_y",&offset_y,"Offset_y/Double_t");
-  geomTree->Branch("Offset_z",&offset_z,"Offset_z/Double_t");
+  geomTree->Branch("GeometryType",geo_type_string,"GeometryType[20]/C");         //example: for std::string data_ : tree->Branch(branchname.c_str(), (void*)data_->c_str(),leafdescription.c_str());
```
Comment (pdeperio) [inline src/WCSimRunAction.cc]: I guess these are what need to be checked with older ROOT versions?

File: src/WCSimVisManager.cc [hunk 3]
```diff
@@ -141,7 +141,8 @@ void WCSimVisManager::RegisterGraphicsSystems () {
     G4cout <<
       "\nYou have successfully chosen to use the following graphics systems."
 	 << G4endl;
-    PrintAvailableGraphicsSystems ();
+    //PrintAvailableGraphicsSystems (); //use this version for Geant4.10.1
```
Comment (pdeperio) [inline src/WCSimVisManager.cc]: able to use precompiler options that automatically grab Geant version from environment?

**Review by pdeperio — CHANGES_REQUESTED** [review]
could you add diagrams and validation plots describing these changes and their impact? (or a link to such slides)

File: include/GdNeutronHPCapture.hh (hunk 1 above)
Comment (kmtsui) [reply include/GdNeutronHPCapture.hh]: resolved in #10

File: src/WCSimRunAction.cc (hunk 2 above)
Comment (kmtsui) [reply src/WCSimRunAction.cc]: resolved in #10

File: src/WCSimVisManager.cc (hunk 3 above)
Comment (kmtsui) [reply src/WCSimVisManager.cc]: resolved in #10

File: src/WCSimConstructPMT.cc [hunk 4]
```diff
@@ -391,7 +391,8 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructPMT(G4String PMTName, G4Str
       // Reflector is a G4PolyCone:
       // Each set of (Z, R, r) defines an edge around the polycone by it's height position (z) inner radius (r) and outer radius (R).
       G4double ReflectorHolderZ[4] = {0*mm, 22.89*mm, 42.04*mm, 47.18*mm};
-      G4double ReflectorHolderR[4] = {39.735*mm, 43.15*mm, 45.89*mm, 46.615*mm};
+      //G4double ReflectorHolderR[4] = {39.735*mm, 43.15*mm, 45.89*mm, 46.615*mm};
+      G4double ReflectorHolderR[4] = {39.733*mm, 43.07*mm, 45.86*mm, 46.615*mm}; // To avoid overflowing the mother cone
```
Comment (kmtsui) [inline src/WCSimConstructPMT.cc:395]: This change makes the reflector not overflowing [solidWCPMT](https://github.com/WCTE/WCSim/blob/288204e52e9654aad377887418797c2b50b8cc19/src/WCSimConstructPMT.cc#L139)

File: src/WCSimConstructMultiPMT.cc [hunk 5]
```diff
@@ -288,7 +288,8 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructMultiPMT(G4String PMTName,
   // origin is center of upward cylinder
   G4LogicalVolume *logic_mPMT_container;
   //necessary to make concentric shells because can Mother can only contain parametrized daughters.
-  G4double innerR_curv_container = vessel_radius_curv - mPMT_outer_material_d - 54.*mm;  //-expose - dist_pmt_vessel
+  //G4double innerR_curv_container = vessel_radius_curv - mPMT_outer_material_d - 54.*mm;  //-expose - dist_pmt_vessel
+  G4double innerR_curv_container = vessel_radius_curv - mPMT_outer_material_d - 59.62*mm;  //updated number with support structure
```
Comment (kmtsui) [inline src/WCSimConstructMultiPMT.cc]: This new number comes from [here](https://github.com/WCTE/WCSim/blob/288204e52e9654aad377887418797c2b50b8cc19/src/WCSimConstructPMT.cc#L84)

File: src/WCSimConstructCylinder.cc [hunk 6]
```diff
@@ -932,6 +932,8 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructCaps(G4int zflip)
   G4double borderAnnulusZ[3] = {(-barrelCellHeight/2.-(WCIDRadius-innerAnnulusRadius))*zflip, 
 								-barrelCellHeight/2.*zflip,
 								barrelCellHeight/2.*zflip};
+  // to check edge case
+  if (WCBarrelPMTOffset < (WCIDRadius-innerAnnulusRadius)) borderAnnulusZ[0] = (-barrelCellHeight/2.-WCBarrelPMTOffset)*zflip;
```
Comment (kmtsui) [inline src/WCSimConstructCylinder.cc:936]: This is to make sure the [barrel ring](https://github.com/WCTE/WCSim/blob/288204e52e9654aad377887418797c2b50b8cc19/src/WCSimConstructCylinder.cc#L939) does not overlap with the [cap](https://github.com/WCTE/WCSim/blob/288204e52e9654aad377887418797c2b50b8cc19/src/WCSimConstructCylinder.cc#L1209)

File: src/WCSimConstructCylinder.cc [hunk 7]
```diff
@@ -1126,7 +1210,8 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructCaps(G4int zflip)
 					   (WCBarrelPMTOffset - (WCIDRadius-innerAnnulusRadius))*zflip,
 					   (WCBarrelPMTOffset - (WCIDRadius-innerAnnulusRadius))*zflip,
 					   WCBarrelPMTOffset*zflip} ;
-
+  // to check edge case
+  if (WCBarrelPMTOffset < (WCIDRadius-innerAnnulusRadius)) { capZ[1] = 0; capZ[2] = 0; }
```
Comment (kmtsui) [inline src/WCSimConstructCylinder.cc:1214]: This is to match with `borderAnnulusZ` in the extreme case

File: src/WCSimConstructCylinder.cc [hunk 8]
```diff
@@ -1230,6 +1315,8 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructCaps(G4int zflip)
   // -------------------------------------------------------------------
   
  G4double capBlackSheetZ[4] = {-WCBlackSheetThickness*zflip, 0., 0., (WCBarrelPMTOffset - (WCIDRadius-innerAnnulusRadius)) *zflip};
+ // to ensure squential z
+ if (WCBarrelPMTOffset - (WCIDRadius-innerAnnulusRadius)<0) capBlackSheetZ[3] = 0;
```
Comment (kmtsui) [inline src/WCSimConstructCylinder.cc:1319]: Again to match with borderAnnulusZ in the extreme case

File: src/WCSimConstructMultiPMT.cc (hunk 5 above)
Comment (pdeperio) [reply src/WCSimConstructMultiPMT.cc]: will this hardcoding in two different places be a problem if someone wants to change it and forgets one place?

Comment (kmtsui) [reply src/WCSimConstructMultiPMT.cc]: changed to directly use `pmtModuleHeight`

File: include/WCSimDetectorConstruction.hh [hunk 9]
```diff
@@ -639,6 +639,7 @@ private:
   G4double vessel_radius_curv;                        // radius of the sphere to determine curvature of cap of pressure vessel
   G4double vessel_radius;                             // radius of the pressure vessel (spherical cap)
   G4double dist_pmt_vessel;                           // distance between glass of pmt and inner radius of pressure vessel (region where water/gel lives)
+  G4double pmtModuleHeight;                           // includes puck and single PMT support, not PMT base. The height of pmt module for solid works design
```
Comment (kmtsui) [inline include/WCSimDetectorConstruction.hh:642]: make this variable available in `WCSimConstructPMT` and `WCSimConstructMultiPMT`

File: src/WCSimConstructCylinder.cc [hunk 10]
```diff
@@ -1035,12 +1037,27 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructCaps(G4int zflip)
...
+                        "WCBarrelBorderCellBlackSheet",
+                        0,0,0);
 
-   G4VPhysicalVolume* physiWCBarrelBorderCellBlackSheet =
+  G4VPhysicalVolume* physiWCBarrelBorderCellBlackSheet =
     new G4PVPlacement(0,
                       G4ThreeVector(0.,0.,0.),
-                      logicWCBarrelCellBlackSheet,
-                      "WCBarrelCellBlackSheet",
+                      logicWCBarrelBorderCellBlackSheet,
```
Comment (kmtsui) [inline src/WCSimConstructCylinder.cc:1059]: The original implementation used the same blacksheet from normal ring cell (those white rectangles enclosing the PMTs), but the top and bottom rings actually has a larger height than normal ring. This makes a region with no blacksheet around the cap.
image
 
The new implementation defines a new blacksheet in the top/bottom ring which covers the missing region.
image

**Comment by pdeperio** [comment]
> validation plots describing their impact

Can you run some photon sims to show the effect of changing mPMT geometry?

And separately blacksheet (e.g. are you able to show that you fixed photons escaping through the gap)?

**Comment by kmtsui** [comment]
To check the effect of mPMT geometry:
Shooting light towards a single mPMT (the central one at top endcap) and check the direct hit rates.

The true hit map looks like:
image

Taking the ratio between original and new geometry (new divided by original). There are differences at the edge but that are of low statistics
image

Sum the hits into each channel (small PMTs). Blue: original, Red: new. The hit rate differences are within statistical fluctuations.
image

**Comment by kmtsui** [comment]
To check the blacksheet effects:
Shoot photons in the gap region.

From event display. 
Original simulation: photons escape the detector. 
image
New simulation: photons got absorbed or reflected. 
image

Turn off photon scattering in water. Shoot photons in the gap and check PMT hits.
Original simulation: no PMT hits
image
New simulation: small amount of photons got reflected and detected.
image

**Review by pdeperio — APPROVED** [review]

**Review by laurenanthony2 — APPROVED** [review]

**Pull request merged**